### PR TITLE
[extractor/iwara] Fix and improve authentication

### LINKED ISSUE #7035: [iwara.tv] Videos with "blocked" tags don't appear in profile list
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that a **supported** site is broken
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Provide a description that is worded well enough to be understood

With the new commit videos with blocked tags can now be downloaded when logged in (this doesn't seem to work with extracted cookies but does work with --username and --password). However when you try to download a profile yt-dlp doesn't find videos with blocked tags, even with the user/pass options. In the output below, yt-dlp only finds one video even though there's 10 videos on the profile when logged in. Seems like yt-dlp isn't logging in when fetching the video list?

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '--username', 'PRIVATE', '--password', 'PRIVATE', 'https://www.iwara.tv/profile/lumymmd']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.05.11.094900 [21b9413cf] (zip)
[debug] Python 3.11.3 (CPython x86_64 64bit) - Linux-6.2.14-200.fc37.x86_64-x86_64-with-glibc2.36 (OpenSSL 3.0.8 7 Feb 2023, glibc 2.36)
[debug] exe versions: ffmpeg 5.1.3 (setts), ffprobe 5.1.3
[debug] Optional libraries: Cryptodome-3.17.0, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1813 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.05.11.094900, Current version: nightly@2023.05.11.094900
Current Build Hash: 68cfb85d91da7da1a1c6f0d01b0f488b89c3275ca89fb893d6e573ba07a30d9d
yt-dlp is up to date (nightly@2023.05.11.094900)
[iwara:user] Extracting URL: https://www.iwara.tv/profile/lumymmd
[iwara:user] lumymmd: Requesting user info
[download] Downloading playlist: Lumy MMD
[iwara:user] lumymmd: Downloading page 0
[iwara:user] Playlist Lumy MMD: Downloading 1 items of 1
[download] Downloading item 1 of 1
[debug] Loading iwara.throw345313 from cache
[debug] Loading iwara.throw345313 from cache
[iwara] Fetching media token
[debug] [iwara] Skipping logging in
[iwara] Extracting URL: https://iwara.tv/video/m3wozcpykc3eap0w
[iwara] m3wozcpykc3eap0w: Downloading JSON metadata
[iwara] m3wozcpykc3eap0w: Downloading JSON metadata
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] m3wozcpykc3eap0w: Downloading 1 format(s): Source
[debug] Invoking http downloader on "https://mikoto.iwara.tv/view?filename=1675934383_M3WOZCpyKC3eap0W_Source.mp4&path=2023%2F02%2F09&expires=1683831320&hash=2e96e96c4da46083979e12832f310fcac7f76c4941404e44156f22b332d75f26"
[download] Destination: Deep Blue Song - Keqing [m3wozcpykc3eap0w].mp4
[download] 100% of  172.25MiB in 00:00:36 at 4.66MiB/s
[download] Finished downloading playlist: Lumy MMD
```

### LINKED ISSUE #7207: Iwara login cache expires
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Provide a description that is worded well enough to be understood

Ever since today, 2th June 2023 (yesterday it was working perfectly), the login API for yt-dlp on Iwara is not working anymore. 

If you try to login using --username --password, all downloads (even those that are not privated/hidden) will fail. 

It will still work on most of the videos if you remove the username/login, but then private/blocked tags videos will not download.

Example: yt-dlp.exe -f b* --username $userhere --password $passwordhere $downloadlink

Result:
[iwara]: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using yt-dlp -U

Using cookies does not work to login on Iwara. Using cookies didn't work on the previous version of yt-dlp either.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', '-f', 'b*', '--username', 'PRIVATE', '--password', 'PRIVATE', 'PRIVATE']
[debug] User config "C:\Users\user\AppData\Roaming\yt-dlp\config.txt": ['-i', '--match-filter', '!is_live', '--download-archive', 'T:/youtube_channel_archiver-master/youtube-dl-archive.txt', '-f', 'bestvideo+bestaudio', '-o', '(%(playlist)s) (%(playlist_id)s)/%(uploader)s (%(uploader_id)s)/%(title)s - [%(resolution)s] [%(id)s] [AGE%(age_limit)s].%(ext)s', '--cookies-from-browser', 'firefox', '--no-check-certificate']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.06.02.150213 [55ed4ff73] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "C:\Users\user\AppData\Roaming\Mozilla\Firefox\Profiles\user\cookies.sqlite"
[Cookies] Extracted 683 cookies from firefox
[debug] Proxy map: {}
[debug] Loaded 1842 extractors
[debug] Loading archive file 'T:/youtube_channel_archiver-master/youtube-dl-archive.txt'
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.06.02.150213, Current version: nightly@2023.06.02.150213
Current Build Hash: 6e916b853695926f9009ae84589467f372ea9990a0cf4ae4a101ae15aab338c3
yt-dlp is up to date (nightly@2023.06.02.150213)
[debug] Loading iwara.PRIVATEUSERNAME from cache
[debug] Loading iwara.PRIVATEUSERNAME from cache
[iwara] Fetching media token
ERROR: [iwara] PRIVATEVIDEOID: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 700, in extract
  File "yt_dlp\extractor\common.py", line 606, in initialize
  File "yt_dlp\extractor\iwara.py", line 185, in _perform_login
  File "yt_dlp\extractor\iwara.py", line 49, in _get_media_token
  File "yt_dlp\extractor\common.py", line 1057, in download_content
  File "yt_dlp\extractor\common.py", line 1021, in download_handle
  File "yt_dlp\extractor\common.py", line 891, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 848, in _request_webpage

  File "yt_dlp\extractor\common.py", line 830, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3776, in urlopen
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default
urllib.error.HTTPError: HTTP Error 403: Forbidden
```

### PR DESCRIPTION
**Pull request by toomyzoom**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This is the same issue in #6671 where access token is needed for some NSFW video. Update to allow user profile and playlist extractors to use the same access token.

Also has fix to renew expired user token in the cache (#7207)

Fixes #7035, Fixes #7207

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🔒📦✅

Improve `IwaraIE` and related extractors. This pull request enhances the authentication and extraction logic for the `IwaraIE`, `IwaraUserIE`, and `IwaraPlaylistIE` extractors, which handle videos from the Iwara website. It also adds a new test case and a helper method for the API headers.

> _We are the extractors of Iwara_
> _We defy the limits of the media_
> _We forge the headers for the API_
> _We renew the tokens before they die_

### Walkthrough
*  Add error handling and debug message for media token request failure ([link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL49-R59))
*  Simplify header construction for API requests by introducing a new helper method `_prepare_auth_header` in `IwaraIE` class ([link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL145-R149), [link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL193-R202))
*  Use `IwaraIE` class as the parent class for `IwaraUserIE` and `IwaraPlaylistIE` classes to inherit the new helper method ([link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL193-R202), [link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL246-R260))
*  Add headers from `_prepare_auth_header` method to API requests for user, playlist, and playlist information extraction in `yt_dlp/extractor/iwara.py` to support private and restricted videos ([link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL228-R242), [link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL263-R278), [link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fL271-R286))
*  Add a new test case for `IwaraUserIE` class to extract videos from a user profile page ([link](https://github.com/yt-dlp/yt-dlp/pull/7137/files?diff=unified&w=0#diff-4a8876f0a8d1dc4bba3feba0da2679799abf8bf4b8080c92fdfb3d069b6a279fR225-R230))

</details>

### CODE REVIEW

File: yt_dlp/extractor/iwara.py [hunk 1]
```diff
@@ -142,11 +145,8 @@ def _extract_formats(self, video_id, fileurl):
 
     def _real_extract(self, url):
         video_id = self._match_id(url)
-        username, password = self._get_login_info()
-        headers = {
-            'Authorization': f'Bearer {self._get_media_token()}',
-        } if username and password else None
-        video_data = self._download_json(f'https://api.iwara.tv/video/{video_id}', video_id, expected_status=lambda x: True, headers=headers)
+        username, _ = self._get_login_info()
+        video_data = self._download_json(f'https://api.iwara.tv/video/{video_id}', video_id, expected_status=lambda x: True, headers=self._prepare_auth_header())
```
Comment (bashonly) [inline yt_dlp/extractor/iwara.py:149]: line too long
```suggestion
        video_data = self._download_json(
            f'https://api.iwara.tv/video/{video_id}', video_id,
            expected_status=lambda x: True, headers=self._prepare_auth_header())
```

File: yt_dlp/extractor/iwara.py [hunk 2]
```diff
@@ -46,14 +46,17 @@ def _get_media_token(self, invalidate=False):
         if not invalidate and self._MEDIATOKEN:
             return self._MEDIATOKEN
```
Comment (bashonly) [inline yt_dlp/extractor/iwara.py:48]:
```suggestion
        if not invalidate and self._MEDIATOKEN:
            return self._MEDIATOKEN
        elif invalidate:
            IwaraBaseIE._MEDIATOKEN = None
```
or else an expired(?) token will still be returned if the exception below is caught

File: yt_dlp/extractor/iwara.py [hunk 3]
```diff
@@ -46,14 +46,17 @@ def _get_media_token(self, invalidate=False):
...
+            IwaraBaseIE._MEDIATOKEN = self._download_json(
+                'https://api.iwara.tv/user/token', None, note='Fetching media token',
+                data=b'',  # Need to have some data here, even if it's empty
+                headers={
+                    'Authorization': f'Bearer {self._get_user_token()}',
+                    'Content-Type': 'application/json'
+                })['accessToken']
+        except ExtractorError as e:
+            if isinstance(e.cause, urllib.error.HTTPError) and e.cause.code == 403:
+                self.write_debug('User Token Expired')
```
Comment (bashonly) [inline yt_dlp/extractor/iwara.py:59]:
```suggestion
        except ExtractorError as e:
            if not isinstance(e.cause, urllib.error.HTTPError) and e.cause.code == 403:
                raise
            self.to_screen('User token has expired')
```

**Review by bashonly — CHANGES_REQUESTED** [review]
Do not subclass `IwaraUserIE` and `IwaraPlaylistIE` from the concrete `IwaraIE` class. Revert to subclassing from `IwaraBaseIE`, and instead move the `_perform_login` and `_prepare_auth_header` methods into the base class.

After moving the methods into the base class, I think `_perform_login` should be changed like so:
```diff
     def _perform_login(self, username, password):
+        if self._USERTOKEN and self._MEDIATOKEN:
+            return
-        if self.cache.load(self._NETRC_MACHINE, username) and self._get_media_token():
+        elif self.cache.load(self._NETRC_MACHINE, username) and self._get_media_token():
             self.write_debug('Skipping logging in')
             return
```
This will prevent a duplicate login/media token request when going from a playlist extractor to the video extractor

File: yt_dlp/extractor/iwara.py [hunk 4]
```diff
@@ -190,8 +190,16 @@ def _perform_login(self, username, password):
         self._get_media_token(True)
         self.cache.store(self._NETRC_MACHINE, username, IwaraBaseIE._USERTOKEN)
 
+    def _prepare_auth_header(self):
+        username, password = self._get_login_info()
+        headers = {
+            'Authorization': f'Bearer {self._get_media_token()}',
+        } if username and password else None
+
+        return headers
```
Comment (bashonly) [inline yt_dlp/extractor/iwara.py:199]:
```suggestion
        if username and password:
            self._get_media_token()

        return {'Authorization': f'Bearer {self._MEDIATOKEN}'} if self._MEDIATOKEN else None
```

**Comment by bashonly** [comment]
You could also check after loading the user token from cache if it is expired:
```diff
diff --git a/yt_dlp/extractor/iwara.py b/yt_dlp/extractor/iwara.py
index 1749dcd1a..64105c39b 100644
--- a/yt_dlp/extractor/iwara.py
+++ b/yt_dlp/extractor/iwara.py
@@ -2,15 +2,18 @@
 import urllib.parse
 import hashlib
 import json
+import time
 
 from .common import InfoExtractor
 from ..utils import (
     ExtractorError,
     OnDemandPagedList,
     int_or_none,
+    jwt_decode_hs256,
     mimetype2ext,
     qualities,
     traverse_obj,
+    try_call,
     unified_timestamp,
 )
 
@@ -27,7 +30,8 @@ def _get_user_token(self, invalidate=False):
 
         username, password = self._get_login_info()
         IwaraBaseIE._USERTOKEN = username and self.cache.load(self._NETRC_MACHINE, username)
-        if not IwaraBaseIE._USERTOKEN or invalidate:
+        expiry = try_call(lambda: jwt_decode_hs256(IwaraBaseIE._USERTOKEN)['exp']) or 0
+        if not IwaraBaseIE._USERTOKEN or invalidate or expiry <= time.time():
             IwaraBaseIE._USERTOKEN = self._download_json(
                 'https://api.iwara.tv/user/login', None, note='Logging in',
                 data=json.dumps({
```

**Comment by toomyzoom** [comment]
> Do not subclass `IwaraUserIE` and `IwaraPlaylistIE` from the concrete `IwaraIE` class. Revert to subclassing from `IwaraBaseIE`, and instead move the `_perform_login` and `_prepare_auth_header` methods into the base class.
> 
> After moving the methods into the base class, I think `_perform_login` should be changed like so:
> 
> ```diff
>      def _perform_login(self, username, password):
> +        if self._USERTOKEN and self._MEDIATOKEN:
> +            return
> -        if self.cache.load(self._NETRC_MACHINE, username) and self._get_media_token():
> +        elif self.cache.load(self._NETRC_MACHINE, username) and self._get_media_token():
>              self.write_debug('Skipping logging in')
>              return
> ```
> 
> This will prevent a duplicate login/media token request when going from a playlist extractor to the video extractor

This should be applied in the last commit.

**Comment by bashonly** [comment]
@toomyzoom I pushed some changes to improve and simplify the authentication code. Now tokens are always invalidated based on expiration time, and the extractor will refresh both tokens as needed. I've tested the changes with and without an account and haven't had any problems. Let me know if you have any problems with it

**Review by bashonly — APPROVED** [review]

**Pull request merged**